### PR TITLE
BigInts should use gmp implementation of gcd and lcm

### PR DESCRIPTION
**Pull request by endSly**
Using gmp methods to calculate gcd and lcm of bigints performance is improved by 4x to 16x.

[Simple benchmark](https://gist.github.com/endSly/4e77f0c6875e8a6777fc860b208afd90).

### CODE REVIEW

File: src/big/lib_gmp.cr [hunk 1]
```diff
@@ -84,6 +84,13 @@ lib LibGMP
   fun cmp_ui = __gmpz_cmp_ui(op1 : MPZ*, op2 : ULong) : Int
   fun cmp_d = __gmpz_cmp_d(op1 : MPZ*, op2 : Double) : Int
 
+  # # Number Theoretic Functions
+
+  fun gcd = __gmpz_gcd(rop : MPZ*, op1 : MPZ*, op2 : MPZ*)
+  fun gcd_ui = __gmpz_gcd_ui(rop : MPZ*, op1 : MPZ*, op2 : MPZ*) : Int
```
Comment (asterite) [inline src/big/lib_gmp.cr]: This binding (and lcm_ui) should have op2 be `ULong`: https://gmplib.org/manual/Number-Theoretic-Functions.html#index-mpz_005fgcd_005fui

File: src/big/big_int.cr [hunk 2]
```diff
@@ -201,6 +201,14 @@ struct BigInt < Int
     BigInt.new { |mpz| LibGMP.pow_ui(mpz, self, other) }
   end
 
+  def gcd(other : Int) : Int
+    BigInt.new { |mpz| LibGMP.gcd(mpz, self, other.to_big_i) }
```
Comment (asterite) [inline src/big/big_int.cr]: We should probably use `gcd_ui` here, probably checking that `other` is not negative (I don't know what should be the result then). The same should be done for `lcm(other : Int)`

**Comment by asterite** [comment]
@endSly Thank you, this is excellent!

I made a few comments, mostly for the overloads that accept an `Int`.

**Comment by endSly** [comment]
Fixed, now operations with ints uses `*_ui` methods. Also added tests for negative numbers.

**Comment by asterite** [comment]
Thank you again! ❤️

File: src/big/big_int.cr [hunk 3]
```diff
@@ -201,6 +201,23 @@ struct BigInt < Int
...
+    result = LibGMP.gcd_ui(nil, self, other.abs.to_u64)
+    result == 0 ? self : result
+  end
+
+  def lcm(other : BigInt) : BigInt
+    BigInt.new { |mpz| LibGMP.lcm(mpz, self, other) }
+  end
+
+  def lcm(other : Int) : BigInt
+    BigInt.new { |mpz| LibGMP.lcm_ui(mpz, self, other.abs.to_u64) }
```
Comment (Sija) [inline src/big/big_int.cr]: What about https://github.com/crystal-lang/crystal/pull/3039#discussion_r71996886 ?
